src/xrpld/app/tx/detail/Credentials.cpp

```diff
@@ -0,0 +1,486 @@
+
+    if (ctx.tx.isFieldPresent(sfIssuer))
+    {
+        // If Issuer present then IssuerPubKey must be the public key of the
+        // regular key for Issuer account. Let's check that it is true.
+
+        auto const sleIssuer = ctx.view.read(keylet::account(issuer));
+        if (!sleIssuer)
+            return tecNO_ISSUER;
+        if (sleIssuer->getAccountID(sfRegularKey) != signer)
+            return tecNO_REGULAR_KEY;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialCreate::doApply()
+{
+    auto const sleOwner = view().peek(keylet::account(account_));
+
+    {
+        STAmount const reserve{view().fees().accountReserve(
+            sleOwner->getFieldU32(sfOwnerCount) + 1)};
+        if (mPriorBalance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    auto const subject = ctx_.tx[~sfSubject].value_or(account_);
+    auto const issuer = ctx_.tx.isFieldPresent(sfIssuer)
+        ? ctx_.tx[sfIssuer]
+        : (ctx_.tx.isFieldPresent(sfIssuerPubKey)
+               ? calcAccountID(PublicKey(ctx_.tx[sfIssuerPubKey]))
+               : account_);
+    auto const credType(ctx_.tx[sfCredentialType]);
+    Keylet const kCred = keylet::credential(subject, issuer, credType);
+    auto const sleCred = std::make_shared<SLE>(kCred);
+
+    auto const optExp = ctx_.tx[~sfExpiration];
+    if (optExp)
+    {
+        std::uint32_t const closeTime =
+            ctx_.view().info().parentCloseTime.time_since_epoch().count();
+
+        if (closeTime > *optExp)
+        {
+            JLOG(j_.trace()) << "Malformed transaction: "
+                                "Expiration time is in the past.";
+            return tecEXPIRED;
+        }
+
+        sleCred->setFieldU32(sfExpiration, ctx_.tx.getFieldU32(sfExpiration));
+    }
+
+    sleCred->setAccountID(sfSubject, subject);
+    sleCred->setAccountID(sfIssuer, issuer);
+    sleCred->setFieldVL(sfCredentialType, credType);
+
+    if (ctx_.tx.isFieldPresent(sfURI))
+        sleCred->setFieldVL(sfURI, ctx_.tx.getFieldVL(sfURI));
+
+    if (!ctx_.tx.isFieldPresent(sfSubject))
+        sleCred->setFieldU32(sfFlags, lsfAccepted);
+
+    view().insert(sleCred);
+    auto const page = view().dirInsert(
+        keylet::ownerDir(account_), kCred, describeOwnerDir(account_));
+
+    JLOG(j_.trace()) << "Adding Credential to owner directory "
+                     << to_string(kCred.key) << ": "
+                     << (page ? "success" : "failure");
+
+    if (!page)
+        return tecDIR_FULL;
+
+    sleCred->setFieldU64(sfOwnerNode, *page);
+
+    adjustOwnerCount(view(), sleOwner, 1, j_);
+    view().update(sleOwner);
+
+    return tesSUCCESS;
+}
+
+// ------- DELETE --------------------------
+NotTEC
+CredentialDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureCredentials))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const optSubj = ctx.tx[~sfSubject];
```

Comment:
please use the full name for these variables - e.g. `optSubject`. With `Subject` I think it's not a huge deal here, but shortening `Issuer` to `Iss` is pretty confusing for me.

Reply:
fixed